cmd/mt-index-analyzer/main.go

```diff
@@ -0,0 +1,199 @@
+package main
+
+import (
+	"flag"
+	"fmt"
+	"os"
+	"regexp"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/grafana/metrictank/conf"
+	"github.com/grafana/metrictank/idx/cassandra"
+	"github.com/grafana/metrictank/idx/memory"
+	"github.com/grafana/metrictank/logger"
+
+	log "github.com/sirupsen/logrus"
+)
+
+var (
+	maxPart      = 512
+	aggregations conf.Aggregations
+	indexRules   conf.IndexRules
+	schemas      conf.Schemas
+
+	aggFile        = "/etc/metrictank/storage-aggregation.conf"
+	indexRulesFile = "/etc/metrictank/index-rules.conf"
+	schemasFile    = "/etc/metrictank/storage-schemas.conf"
+
+	totalPointsInTank  uint64
+	totalPointsInStore uint64
+	totalSeries        uint64
+
+	dist    sync.Mutex
+	irDist  []int
+	aggDist []int
+)
+
+func init() {
+	formatter := &logger.TextFormatter{}
+	formatter.TimestampFormat = "2006-01-02 15:04:05.000"
+	log.SetFormatter(formatter)
+	log.SetLevel(log.InfoLevel)
+}
+
+func perror(err error) {
+	if err != nil {
+		log.Fatal(err.Error())
+	}
+}
+
+func main() {
+	fl := cassandra.ConfigSetup()
+	fl.StringVar(&aggFile, "aggregations-file", aggFile, "path to storage-aggregation.conf file")
+	fl.StringVar(&indexRulesFile, "rules-file", indexRulesFile, "path to index-rules.conf file")
+	fl.StringVar(&schemasFile, "schemas-file", schemasFile, "path to storage-schemas.conf file")
+	fl.IntVar(&maxPart, "max-partition", maxPart, "id of the highest partition to look for")
+
+	flag.Usage = func() {
+		fmt.Println("mt-index-analyzer")
+		fmt.Println("")
+		fmt.Println("analyzes the contents of your index")
+		fmt.Println("limitations:")
+		fmt.Println(" - does not take into account series churn. more specifically, assumes each series has data for its entire defined retention")
+		fmt.Println(" - assumes chunks are filled corresponding to the interval, doesn't account for sparse data or too-high resolution data")
+		fl.PrintDefaults()
+	}
+	if len(os.Args) == 2 && (os.Args[1] == "-h" || os.Args[1] == "--help") {
+		flag.Usage()
+		os.Exit(0)
+	}
+
+	fl.Parse(os.Args[1:])
+	cassandra.CliConfig.Enabled = true
+	cassandra.ConfigProcess()
+	idx := cassandra.New(cassandra.CliConfig)
+	err := idx.InitBare()
+	perror(err)
+
+	// for loading the index, to not drop any entries
+	memory.IndexRules = conf.IndexRules{
+		Rules: nil,
+		Default: conf.IndexRule{
+			Name:     "default",
+			Pattern:  regexp.MustCompile(""),
+			MaxStale: 0,
```

Comment:
I think we actually probably want to set this to a lower value.  Generating a report that includes series that are "stale" will lead to wildly inaccurate numbers.

We mostly want to generate the report based on "Active Series".   Series churn does not really impact the amount of data being stored, provided it is not too frequent.  eg, when looking at the metrics storage for a single application running in k8s that has 10 pods each sending 100 metrics.  If the pods are replaced every 2days, the amount of data stored wont really change.  We will always be writing 1000 chunks.

Reply:
yes, to find the sweet spot here we have to trade off:
1) a too aggressive value will exclude series that have temporarily not received data.
2) a too loose value may accidentally incorporate churn events, e.g. multiple series that were supposed to count for the same single "logical" series that cycled onto a new name

I think i'll just make this a flag with a default of 1h or so.